Demultiplex cell-hashing/nuclei-hashing data using demuxEM or perform CITE-Seq analysis
---------------------------------------------------------------------------------------

Follow the steps below to run **scCloud** for cell-hashing/nuclei-hashing/CITE-Seq data on FireCloud.

#. Create a sample sheet, **sample_sheet_hashing.csv**, which describes the metadata for each pair of RNA and ADT data. The sample sheet should contain 4 columns --- *OUTNAME*, *RNA*, *ADT*, and *TYPE*. *OUTNAME* is the output name for one pair of RNA and ADT data. *RNA* is the raw gene count matrix generated by *cellranger count* (e.g. ``raw_gene_bc_matrices_h5.h5``). *ADT* is the ADT count matrix generated by *scCloud_adt* (e.g. ``sample_ADT.csv``). *TYPE* is the assay type, which can be *cell-hashing*, *nuclei-hashing*, or *cite-seq*.

	Example::

		OUTNAME,RNA,ADT,TYPE
		sample_1,gs://fc-e0000000-0000-0000-0000-000000000000/my_dir/sample_1/raw_gene_bc_matrices_h5.h5,gs://fc-e0000000-0000-0000-0000-000000000000/my_dir/sample_1_ADT/sample_1_ADT.csv,cell-hashing
		sample_2,gs://fc-e0000000-0000-0000-0000-000000000000/my_dir/sample_2/raw_gene_bc_matrices_h5.h5,gs://fc-e0000000-0000-0000-0000-000000000000/my_dir/sample_2_ADT/sample_2_ADT.csv,nuclei-hashing

#. Upload your sample sheet to the workspace.  

	Example::
	
		gsutil cp /foo/bar/projects/my_sample_sheet_hashing.csv gs://fc-e0000000-0000-0000-0000-000000000000/

#. Import **scCloud_hashing_cite_seq** method.

	In FireCloud, select the "Method Configurations" tab then click "Import Configuration". Click "Import From Method Repository". Type **scCloud/scCloud_hashing_cite_seq**.

#. Uncheck "Configure inputs/outputs using the Workspace Data Model"

---------------------------------

scCloud_hashing_cite_seq inputs:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. list-table::
	:widths: 5 20 10 5
	:header-rows: 1

	* - Name
	  - Description
	  - Example
	  - Default
	* - **input_sample_sheet**
	  - Input CSV file describing metadata of RNA and ADT data pairing
	  - "my_sample_sheet_hashing.csv"
	  - 
	* - **output_directory**
	  - This is the output directory (gs url + path) for all results. There will be one folder per RNA-ADT data pair under this directory
	  - "gs://fc-e0000000-0000-0000-0000-000000000000/my_demux_dir"
	  - 
	* - genome
	  - Reference genome name. If not provided, scCloud will infer the genome name from data
	  - "GRCh38"
	  - 
	* - demuxEM_min_num_genes
	  - demuxEM parameter. Only demultiplex cells/nuclei with at least <demuxEM_min_num_genes> expressed genes
	  - 200
	  - 100
	* - demuxEM_max_background_probability
	  - demuxEM parameter. Any cell/nucleus with no less than <demuxEM_max_background_probability> background probability will be marked as unknown
	  - 0.8
	  - 0.8
	* - demuxEM_random_state
	  - demuxEM parameter. The random seed used in the KMeans algorithm to separate empty ADT droplets from others
	  - 0
	  - 0
	* - demuxEM_generate_diagnostic_plots
	  - demuxEM parameter. If generate a series of diagnostic plots, including the background/signal between HTO counts, estimated background probabilities, HTO distributions of cells and non-cells etc
	  - true
	  - true
	* - demuxEM_generate_gender_plot
	  - demuxEM parameter. If generate violin plots using gender-specific genes (e.g. Xist). <demuxEM_generate_gender_plot> is a comma-separated list of gene names
	  - "XIST"
	  - 
	* - num_cpu
	  - Number of cpus per scCloud_hashing_cite_seq job
	  - 8
	  - 8
	* - memory
	  - Memory size in GB
	  - 10
	  - 10
	* - disk_space
	  - Total disk space
	  - 20
	  - 20
	* - preemptible
	  - Number of preemptible tries
	  - 2
	  - 2

---------------------------------

scCloud_hashing_cite_seq outputs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

See the table below for important *scCloud_hashing_cite_seq* outputs:

.. list-table::
	:widths: 5 5 10
	:header-rows: 1

	* - Name
	  - Type
	  - Description
	* - output_folder
	  - Array[String]
	  - A list of google bucket urls containing results for every RNA-ADT data pairs.

In the output folder of each RNA-ADT data pair, you can find the following files:

.. list-table::
	:widths: 5 10
	:header-rows: 1

	* - Name
	  - Description
	* - output_name_demux.h5ad
	  - Demultiplexed RNA count matrix in h5ad format.
	* - output_name_demux_10x.h5
	  - RNA expression matrix with demultiplexed sample identities in 10x's hdf5 format.
	* - output_name_ADTs.h5ad
	  - Antibody tag matrix in h5ad format.
	* - output_name.ambient_hashtag.hist.png
	  - Optional output. A histogram plot depicting hashtag distributions of empty droplets and non-empty droplets.
	* - output_name.background_probabilities.bar.png
	  - Optional output. A bar plot visualizing the estimated hashtag background probability distribution.
	* - output_name.real_content.hist.png
	  - Optional output. A histogram plot depicting hashtag distributions of not-real-cells and real-cells as defined by total number of expressed genes in the RNA assay.
	* - output_name.rna_demux.hist.png
	  - Optional output. A histogram plot depicting RNA UMI distribution for singlets, doublets and unknown cells.
	* - output_name.gene_name.violin.png
	  - Optional outputs. Violin plots depicting gender-specific gene expression across samples. We can have multiple plots if a gene list is provided in input 'demuxEM_generate_gender_plot'.

---------------------------------

Load demultiplexing results into ``Python`` and ``R``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To load demultiplexing results into ``Python``, you need to install Python package ``anndata`` first. Then follow the codes below::

	import anndata
	data = anndata.read_h5ad('output_name_demux.h5ad')

Once you load the data object, you can find predicted droplet types (singlet/doublet/unknown) in ``data.obs['demux_type']``. You can find predicted sample assignments in ``data.obs['assignment']``. You can find estimated sample fractions (sample1, sample2, ..., samplen, background) for each droplet in ``data.obsm['raw_probs']``.

To load the results into ``R``, you need to install R package ``reticulate`` in addition to Python package ``anndata``. Then follow the codes below::

	library(reticulate)
	ad <- import("anndata", convert = FALSE)
	data <- ad$read_h5ad("output_name_demux.h5ad")

Results are in ``data$obs['demux_type']``, ``data$obs['assignment']``, and ``data$obsm['raw_probs']``.

---------------------------------

CITE-Seq data analysis
----------------------

To be done.
